flink-end-to-end-tests/flink-end-to-end-tests-common/src/main/java/org/apache/flink/tests/util/flink/container/FlinkContainers.java

```diff
@@ -0,0 +1,254 @@
+
+import org.junit.jupiter.api.extension.AfterAllCallback;
+import org.junit.jupiter.api.extension.BeforeAllCallback;
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.testcontainers.containers.Container;
+import org.testcontainers.containers.GenericContainer;
+import org.testcontainers.utility.MountableFile;
+
+import javax.annotation.Nullable;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.TimeoutException;
+
+import static org.apache.flink.util.Preconditions.checkState;
+
+/** A Flink cluster running JM and TM on different containers. */
+public class FlinkContainers implements BeforeAllCallback, AfterAllCallback {
+    private static final Logger LOG = LoggerFactory.getLogger(FlinkContainers.class);
+
+    // Hostname of components within container network
+    public static final String JOB_MANAGER_HOSTNAME = "jobmanager";
+    public static final String TASK_MANAGER_HOSTNAME_PREFIX = "taskmanager-";
+    public static final String ZOOKEEPER_HOSTNAME = "zookeeper";
+
+    // Directories for storing states
+    public static final Path CHECKPOINT_PATH = Paths.get("/flink/checkpoint");
+    public static final Path HA_STORAGE_PATH = Paths.get("/flink/recovery");
+
+    // Default timeout of operations
+    public static final Duration DEFAULT_TIMEOUT = Duration.ofSeconds(30);
+
+    private final GenericContainer<?> jobManager;
+    private final List<GenericContainer<?>> taskManagers;
+    private final GenericContainer<?> haService;
+    private final Configuration conf;
+
+    private RestClusterClient<StandaloneClusterId> restClusterClient;
+    private boolean isStarted;
+
+    /** Creates a builder for {@link FlinkContainers}. */
+    public static FlinkContainersBuilder builder() {
+        return new FlinkContainersBuilder();
+    }
+
+    FlinkContainers(
+            GenericContainer<?> jobManager,
+            List<GenericContainer<?>> taskManagers,
+            @Nullable GenericContainer<?> haService,
+            Configuration conf) {
+        this.jobManager = jobManager;
+        this.taskManagers = taskManagers;
+        this.haService = haService;
+        this.conf = conf;
+    }
+
+    /** Starts all containers. */
+    public void start() throws Exception {
+        if (haService != null) {
+            LOG.debug("Starting HA service container");
+            this.haService.start();
+        }
+        LOG.debug("Starting JobManager container");
+        this.jobManager.start();
+        LOG.debug("Starting TaskManager containers");
+        this.taskManagers.parallelStream().forEach(GenericContainer::start);
+        LOG.debug("Creating REST cluster client");
+        this.restClusterClient = createClusterClient();
```

Comment:
I think this call can become flaky because it is not guaranteed that the jobmanager is reachable when we try to build the cluster client. We need a similar readiness check to what we use for the taskmanagers.